python/ray/serve/_private/controller.py

```diff
@@ -1032,6 +1046,63 @@ def _get_logging_config(self) -> Tuple:
         return self.system_logging_config, log_file_path
 
 
+def calculate_scale_direction(
+    new_config: ServeDeploySchema,
+    all_app_statuses: Dict[str, ApplicationStatusInfo],
+    curr_target_capacity: Optional[float],
+    curr_scale_direction: Optional[float],
+) -> Optional[TargetCapacityScaleDirection]:
+    next_scale_dirction = None
+
+    if len(all_app_statuses) == 0 and new_config.target_capacity is not None:
+        next_scale_dirction = TargetCapacityScaleDirection.UP
+
+    if live_applications_match_config(new_config, all_app_statuses):
+        next_target_capacity = new_config.target_capacity
+
+        if curr_target_capacity == next_target_capacity:
+            next_scale_dirction = curr_scale_direction
+        elif curr_target_capacity is None and next_target_capacity is not None:
+            next_scale_dirction = TargetCapacityScaleDirection.DOWN
+        elif curr_target_capacity is None and next_target_capacity is not None:
+            next_scale_dirction = None
+        elif curr_target_capacity < next_target_capacity:
+            next_scale_dirction = TargetCapacityScaleDirection.UP
+        else:
+            assert curr_target_capacity > next_target_capacity
+            next_scale_dirction = TargetCapacityScaleDirection.DOWN
+    else:
+        next_scale_dirction = curr_scale_direction
+
+    if next_scale_dirction != curr_scale_direction:
+        if isinstance(next_scale_dirction, TargetCapacityScaleDirection):
+            logger.info(
+                f"Target capacity scaling {next_scale_dirction.value.lower()} "
+                f"from {curr_target_capacity} to {new_config.target_capacity}."
+            )
+        else:
+            assert next_scale_dirction is None
+            logger.info(f"Target capacity entering 100% at steady state.")  # noqa: F541
+
+    return next_scale_dirction
+
+
+def live_applications_match_config(
```

Comment:
That's a good point– I tried doing that initially, but it makes it really hard to test this logic. I would need to mock out an `application_state_manager` and then somehow populate its internal data-structures with entirely fake apps, which requires mocking out `ray.get` methods.

E.g. [here's an example unit test](https://github.com/ray-project/ray/blob/25bee340cc9cf751b3212e8b0d8b618b21869399/python/ray/serve/tests/unit/test_application_state.py#L413-L452) that deploys 1 config to the `application_state_manager`. I would need to do that in order to populate the manager, and then I would have to write the actual testing logic afterwards.

I could move these functions into the `ServeController` as a static methods. They would still be easier to test, and it might be more organized. What do you think?

Reply:
Yep, with the new implementation, static method is fine :)

Reply:
Actually, a static method is a bit messy. Since the controller is a Ray actor, we can't access its static methods directly. I'll leave the functions as-is. Let me know if you think I should change it.

Reply:
Leave as is is good :)